modules/dragster/src/main/ts/ephox/dragster/datatransfer/DataTransfer.ts

```diff
@@ -0,0 +1,131 @@
+
+  const dataTransfer: DataTransfer = {
+    get dropEffect() {
+      return dropEffect;
+    },
+
+    set dropEffect(effect: DataTransfer['dropEffect']) {
+      if (Arr.contains(validDropEffects, effect)) {
+        dropEffect = effect;
+      }
+    },
+
+    get effectAllowed() {
+      return effectAllowed;
+    },
+
+    set effectAllowed(allowed: DataTransfer['effectAllowed']) {
+      // TINY-9601: Only allow setting effectAllowed to a valid value in a dragstart event
+      // https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/effectAllowed
+      if (isInDragStartEvent(dataTransfer) && Arr.contains(validEffectAlloweds, allowed)) {
+        effectAllowed = allowed;
+      }
+    },
+
+    get items() {
+      return normalizeItems(dataTransfer, dataTransferImpl.items);
+    },
+
+    get files() {
+      if (isInProtectedMode(dataTransfer)) {
+        return createEmptyFileList();
+      } else {
+        return dataTransferImpl.files;
+      }
+    },
+
+    get types() {
+      return dataTransferImpl.types;
+    },
+
+    setDragImage: (image: Element, x: number, y: number): void => {
+      if (isInReadWriteMode(dataTransfer)) {
+        setDragImage(dataTransfer, { image, x, y });
+        dataTransferImpl.setDragImage(image, x, y);
+      }
+    },
+
+    getData: (format: string): string => {
+      if (isInProtectedMode(dataTransfer)) {
+        return '';
+      } else {
+        return dataTransferImpl.getData(format);
+      }
+    },
+
+    setData: (format: string, data: string): void => {
+      if (isInReadWriteMode(dataTransfer)) {
+        dataTransferImpl.setData(format, data);
+      }
+    },
+
+    clearData: (format?: string | undefined): void => {
+      if (isInReadWriteMode(dataTransfer)) {
+        dataTransferImpl.clearData(format);
+      }
+    }
+  };
+
+  setReadWriteMode(dataTransfer);
+
+  return dataTransfer;
+};
+
+const cloneDataTransfer = (original: DataTransfer): DataTransfer => {
+  // Create new DataTransfer object to ensure scope is not shared between original and clone
+  const clone = createDataTransfer();
+
+  // Store original mode and set to read-only to copy data
+  const originalMode = getMode(original);
+  setReadOnlyMode(original);
+
+  // Set clone event to dragstart to ensure effectAllowed can be set
+  setDragstartEvent(clone);
+
+  clone.dropEffect = original.dropEffect;
+  clone.effectAllowed = original.effectAllowed;
+  getDragImage(original).each((imageData) => clone.setDragImage(imageData.image, imageData.x, imageData.y));
+
+  // Copy string data
+  Arr.each(original.types, (type) => {
+    if (type !== 'Files') {
+      clone.setData(type, original.getData(type));
+    }
+  });
+
+  // Copy files
+  Arr.each(original.files, (file) => clone.items.add(file));
+
+  // Set event type
+  getEvent(original).each((type) => {
+    setEvent(clone, type);
+  });
+
+  // Set mode
+  originalMode.each((mode) => {
+    setMode(original, mode);
```

Comment:
Is this correct why is it setting the mode on the original?

Reply:
This is because it was set to read-only earlier for data copying purposes, so now we want to reset the mode of the original to what it was before https://github.com/tinymce/tinymce/blob/e2062fd1a8bef54a3d99e597ec5875fc586fe2b5/modules/dragster/src/main/ts/ephox/dragster/datatransfer/DataTransfer.ts#L98

Reply:
But shouldn't you always be able to read data regardless of what mode it's in? My memory on the spec if fuzzy.